### PR TITLE
Fix thread sanitizer

### PR DESCRIPTION
**Pull request by bansan85**
See drogonframework/drogon#1207

### CODE REVIEW

**Comment by bansan85** [comment]
Damned, it looks like it's not the good fix:

```
1: WARNING: ThreadSanitizer: data race (pid=8910)
1:   Write of size 8 at 0x7ba0000000a0 by thread T5:
1:     #0 close <null> (test_calc_meyerhof+0x467333)
1:     #1 trantor::EventLoop::~EventLoop() /home/runner/work/jessica/jessica/build/_deps/drogon-src/trantor/trantor/net/EventLoop.cc:141:5 (test_calc_meyerhof+0xc9eb4f)
1:     #2 __call_tls_dtors <null> (libc.so.6+0x45dfe)
1: 
1:   Previous read of size 8 at 0x7ba0000000a0 by thread T4:
1:     #0 write <null> (test_calc_meyerhof+0x46d941)
1:     #1 trantor::EventLoop::wakeup() /home/runner/work/jessica/jessica/build/_deps/drogon-src/trantor/trantor/net/EventLoop.cc:305:15 (test_calc_meyerhof+0xc9eeab)
1:     #2 trantor::EventLoop::quit() /home/runner/work/jessica/jessica/build/_deps/drogon-src/trantor/trantor/net/EventLoop.cc:170:9 (test_calc_meyerhof+0xc9ec11)
1:     #3 trantor::EventLoopThread::~EventLoopThread() /home/runner/work/jessica/jessica/build/_deps/drogon-src/trantor/trantor/net/EventLoopThread.cc:37:15 (test_calc_meyerhof+0xca478f)
```

**Comment by bansan85** [comment]
There is 2 data races in `EventLoopThread::~EventLoopThread` when destroying software.

First problem:
  - read of `loop_` in EventLoopThread::~EventLoopThread
  - write of `loop_` in EventLoopThread::loopFuncs

Solution : use `mutex` to secure these 2 operations.

Second problem: `thread_local static EventLoop loop` in `EventLoopThread::loopFuncs`.

This global variable may be destroyed when closing software. But at the same time, `loop_` may still in use in `EventLoopThread::~EventLoopThread` in another thread while running `loop_->quit()`.

Solution is to replace `thread_local static EventLoop` by `thread_local static std::shared_ptr<EventLoop>`.

File: trantor/net/EventLoopThread.h [hunk 1]
```diff
@@ -59,10 +59,13 @@ class TRANTOR_EXPORT EventLoopThread : NonCopyable
     void run();
 
   private:
-    EventLoop *loop_;
+    // With C++20, use std::atomic<std::shared_ptr<EventLoop>>
+    std::shared_ptr<EventLoop> loop_;
+    std::mutex m_loop_;
```
Comment (hwc0919) [inline trantor/net/EventLoopThread.h:64]: Use a name in consistence with drogon/trantor 's naming conventions. Such as `mutex_` or `loopMutex_`.

**Review by hwc0919 — CHANGES_REQUESTED** [review]

**Comment by hwc0919** [comment]
A good one. Should be able to solve some exit issue.

**Review by an-tao — APPROVED** [review]

File: trantor/net/EventLoopThread.h (hunk 1 above)
Comment (bansan85) [reply trantor/net/EventLoopThread.h]: Done

**Review by hwc0919 — APPROVED** [review]

**Comment by bansan85** [comment]
Nice 😄 Thanks

**Pull request merged**